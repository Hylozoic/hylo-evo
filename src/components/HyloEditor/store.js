import { createSelector } from 'reselect'
import { fromJS } from 'immutable'
import { defaultSuggestionsFilter } from 'draft-js-mention-plugin'
import sampleMentions from './sampleMentions'

const FIND_MENTIONS = 'hyloEditor/FIND_MENTIONS'
const CLEAR_MENTIONS = 'hyloEditor/CLEAR_MENTIONS'

// Action Creators

export function findMentions (searchText) {
  return {
    type: FIND_MENTIONS,
    payload: { searchText }
  }
}

export function clearMentions (searchText) {
  return { type: CLEAR_MENTIONS }
}

// Reducer

export default function reducer (state = {mentionResults: fromJS([])}, action) {
  const { error, type, payload } = action
  if (error) return state

  switch (type) {
    case FIND_MENTIONS:
      return {...state, mentionResults: defaultSuggestionsFilter(payload.searchText, fromJS(sampleMentions))}
    case CLEAR_MENTIONS:
      return {...state, mentionResults: fromJS([])}
    default:
      return state
  }
}

// Selectors

export const getMentionResults = createSelector(
  (state) => state.hyloEditor,
  (state, props) => state.mentionResults
)
