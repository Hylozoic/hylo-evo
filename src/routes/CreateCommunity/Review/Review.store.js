import orm from 'store/models'
import { createSelector as ormCreateSelector } from 'redux-orm'
import { AnalyticsEvents } from 'hylo-utils/constants'

export const MODULE_NAME = `Review`
export const CREATE_COMMUNITY = `${MODULE_NAME}/CREATE_COMMUNITY`

export function createCommunity (name, slug, networkId) {
  const data = {
    name,
    slug
  }

  if (networkId) {
    data.networkId = networkId
  }

  return {
    type: CREATE_COMMUNITY,
    graphql: {
      query: `mutation ($data: CommunityInput) {
        createCommunity(data: $data) {
          id
          hasModeratorRole
          community {
            id
            name
            slug
            network {
              id
            }
          }
        }
      }
      `,
      variables: {
        data
      }
    },
    meta: {
      extractModel: 'Membership',
      ...data,
      analytics: AnalyticsEvents.COMMUNITY_CREATED
    }
  }
}

export const getNetwork = ormCreateSelector(
  orm,
  state => state.orm,
  (state, { networkId }) => networkId,
  (session, networkId) => {
    const network = session.Network.safeGet({ id: networkId })
    if (network) {
      return network
    }
    return null
  }
)
