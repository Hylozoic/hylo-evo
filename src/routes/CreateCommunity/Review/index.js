import component from './Review'
import connector from './Review.connector'
export default connector(component)
