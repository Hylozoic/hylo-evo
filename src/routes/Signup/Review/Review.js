import React, { Component } from 'react'
import { get } from 'lodash/fp'
import LeftSidebar from '../LeftSidebar'
import SignupModalFooter from '../SignupModalFooter'
import UploadImageSection from '../UploadImageSection'

import '../Signup.scss'

export default class AddLocation extends Component {
  constructor () {
    super()
    this.state = {
      edits: {
        name: null,
        email: null,
        location: null
      },
      readOnly: {
        name: true,
        email: true,
        location: true
      }
    }
  }
  handleInputChange = (event, name) => {
    const value = event.target.value
    this.setState({
      edits: {
        ...this.state.edits,
        [name]: value
      }
    })
  }

  makeEditable = (event, name) => {
    this.setState({
      readOnly: {
        ...this.state.edits,
        [name]: false
      }
    })
  }
  submit = () => {
    const { edits } = this.state
    Object.keys(edits).forEach((key) => (edits[key] == null) && delete edits[key])
    const changes = Object.assign(edits, {settings: {signupInProgress: false}})
    this.props.updateUserSettings(changes)
    this.props.goToNextStep()
  }

  previous = () => {
    this.props.goToPreviousStep()
  }

  updateSettingDirectly = (key, setChanged) => value => {
    const { edits, changed } = this.state
    // setChanged && setConfirm('You have unsaved changes, are you sure you want to leave?')
    this.setState({
      changed: setChanged ? true : changed,
      edits: {
        ...edits,
        [key]: value
      }
    })
  }

  componentWillMount = () => {
    const { currentUser } = this.props
    if (currentUser && !get('settings.signupInProgress', currentUser)) this.props.goBack()
  }

  componentDidMount = () => {
    this.props.fetchMySkills()
  }

  getValue = (field) => {
    return this.state.edits[field] || get(field, this.props.currentUser)
  }

  render () {
    const currentAvatarUrl = this.state.edits.avatarUrl
    const { currentUser, uploadImagePending } = this.props
    return <div styleName='flex-wrapper'>
      <LeftSidebar
        header='Everything looking good?'
        body='You can always come back and change your details at any time.'
      />
      <div styleName='panel'>
        <span styleName='white-text step-count'>STEP 4/4</span>
        <br />
        <div styleName='center'>
          {currentUser && <UploadImageSection
            avatarUrl={currentAvatarUrl}
            updateSettingDirectly={this.updateSettingDirectly}
            currentUser={currentUser}
            loading={uploadImagePending}
          />}
        </div>
        <div styleName='final-edit'>
          <div styleName='three-column-input gray-bottom-border'>
            <div styleName='left-input-column'>
              <span styleName='text-opacity'>YOUR NAME</span>
            </div>
            <div styleName='center-input-column'>
              <input
                styleName='signup-input review-input-padding'
                onChange={(e) => this.handleInputChange(e, 'name')}
                onKeyPress={event => {
                  if (event.key === 'Enter') {
                    this.submit()
                  }
                }}
                autoFocus
                value={this.getValue('name')}
                readOnly={this.state.readOnly.name}
              />
            </div>
            <div styleName='right-input-column'>
              <span styleName='edit-button text-opacity' onClick={() => this.makeEditable('name')}>Edit</span>
            </div>
          </div>
          <div styleName='three-column-input gray-bottom-border'>
            <div styleName='left-input-column'>
              <span styleName='text-opacity'>YOUR EMAIL</span>
            </div>
            <div styleName='center-input-column'>
              <input
                styleName='signup-input review-input-padding'
                onChange={(e) => this.handleInputChange(e, 'email')}
                onKeyPress={event => {
                  if (event.key === 'Enter') {
                    this.submit()
                    this.props.goToNextStep()
                  }
                }}
                autoFocus
                value={this.getValue('email')}
                readOnly={this.state.readOnly.email}
              />
            </div>
            <div styleName='right-input-column'>
              <span styleName='edit-button text-opacity' onClick={() => this.makeEditable('email')}>Edit</span>
            </div>
          </div>
          <div styleName='three-column-input gray-bottom-border'>
            <div styleName='left-input-column'>
              <span styleName='text-opacity'>LOCATION</span>
            </div>
            <div styleName='center-input-column'>
              <input
                styleName='signup-input review-input-padding'
                onChange={(e) => this.handleInputChange(e, 'location')}
                onKeyPress={event => {
                  if (event.key === 'Enter') {
                    this.submit()
                    this.props.goToNextStep()
                  }
                }}
                autoFocus
                value={this.getValue('location')}
                readOnly={this.state.readOnly.location}
              />
            </div>
            <div styleName='right-input-column'>
              <span styleName='edit-button text-opacity' onClick={() => this.makeEditable('email')}>Edit</span>
            </div>
          </div>
          <div styleName='three-column-input gray-bottom-border'>
            <div styleName='left-input-column'>
              <span styleName='text-opacity'>SKILLS</span>
            </div>
            <div styleName='center-input-column-scrollable'>
              {this.props.skills.map((skill, index) =>
                <div key={index}>
                  <Pill skill={skill} handlerArg={'name'} />
                </div>
              )}
            </div>
            <div styleName='right-input-column'>
              <span styleName='edit-button text-opacity' onClick={this.props.goBack}>Edit</span>
            </div>
          </div>
        </div>
        <div>
          <SignupModalFooter submit={this.submit} previous={this.previous} showPrevious={false} continueText={"Let's do this!"} />
        </div>
      </div>
    </div>
  }
}

export function Pill ({skill}) {
  return <span styleName='review-skill'>
    {skill.name}
  </span>
}
