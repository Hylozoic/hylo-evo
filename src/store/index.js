import { createStore } from 'redux'
import createMiddleware from './middleware'
import orm from './models'
import rootReducer, { combinedReducers } from './reducers'

export function getEmptyState () {
  return combinedReducers({
    orm: orm.getEmptyState()
  }, {type: ''})
}

export default function (history, req) {
  return createStore(rootReducer, getEmptyState(), createMiddleware(history, req))
}
