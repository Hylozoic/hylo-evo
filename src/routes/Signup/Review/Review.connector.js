import { connect } from 'react-redux'
import { push, goBack } from 'react-router-redux'
import getMe from 'store/selectors/getMe'
import { UPLOAD_ATTACHMENT } from 'store/constants'
import { updateUserSettings } from 'store/actions/updateUserSettings'
import fetchMySkills from 'store/actions/fetchMySkills'
import getMySkills from 'store/selectors/getMySkills'

export function mapStateToProps (state, props) {
  const uploadImagePending = state.pending[UPLOAD_ATTACHMENT]
  return {
    currentUser: getMe(state),
    uploadImagePending,
    skills: getMySkills(state)
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    updateUserSettings: (changes) => dispatch(updateUserSettings(changes)),
    goToNextStep: () => dispatch(push('/')),
    goToPreviousStep: () => dispatch(push('/signup/add-skills')),
    goBack: () => dispatch(goBack()),
    fetchMySkills: () => dispatch(fetchMySkills())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
