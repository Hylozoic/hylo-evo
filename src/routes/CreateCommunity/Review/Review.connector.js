import { get } from 'lodash/fp'
import { connect } from 'react-redux'
import { push } from 'react-router-redux'
import getMe from 'store/selectors/getMe'
import updateUserSettings from 'store/actions/updateUserSettings'
import { addCommunityName, addCommunityDomain, fetchCommunityExists } from '../CreateCommunity.store'
import { createCommunity, getNetwork } from './Review.store'

export function mapStateToProps (state, props) {
  const communityNetworkId = get('networkId', state.CreateCommunity)
  const network = getNetwork(state, {networkId: communityNetworkId})

  return {
    currentUser: getMe(state),
    communityDomain: get('domain', state.CreateCommunity),
    communityName: get('name', state.CreateCommunity),
    communityPrivacy: get('privacy', state.CreateCommunity),
    communityNetworkId,
    communityDomainExists: get('domainExists', state.CreateCommunity),
    networkName: get('name', network)
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
    goToCommunity: (communityDomain) => dispatch(push(`/c/${communityDomain}`)),
    updateUserSettings: (changes) => dispatch(updateUserSettings(changes)),
    clearNameFromCreateCommunity: () => dispatch(addCommunityName(null)),
    clearDomainFromCreateCommunity: () => dispatch(addCommunityDomain(null)),
    createCommunity: (name, slug, networkId) => dispatch(createCommunity(name, slug, networkId)),
    goToPrivacyStep: () => dispatch(push('/create-community/privacy')),
    goHome: () => dispatch(push('/')),
    fetchCommunityExists: (slug) => dispatch(fetchCommunityExists(slug))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
