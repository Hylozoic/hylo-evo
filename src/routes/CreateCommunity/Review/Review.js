import React, { Component } from 'react'
import '../CreateCommunity.scss'
import ModalSidebar from 'components/ModalSidebar'
import TextInput from 'components/TextInput'
import { hyloNameWhiteBackground, groovingAxolotl } from 'util/assets'
import { bgImageStyle } from 'util/index'
import ModalFooter from 'components/ModalFooter'
import { find } from 'lodash'
import { get } from 'lodash/fp'
import {
  slugValidatorRegex,
  formatDomainWithUrl,
  removeUrlFromDomain,
  invalidSlugMessage
} from '../util'

export default class Review extends Component {
  constructor () {
    super()

    this.state = {
      readOnly: {
        name: true,
        email: true,
        communityName: true,
        communityDomain: true
      },
      edits: {
        name: '',
        email: '',
        communityName: '',
        communityDomain: '',
        communityPrivacy: null,
        changed: false
      }
    }
  }

  editHandler = (name) => {
    this.setState({
      readOnly: {
        ...this.state.readOnly,
        [name]: false
      }
    })
    this[name].select()
  }

  handleInputChange = (event, name) => {
    let value = event.target.value
    if (name === 'communityDomain') {
      value = removeUrlFromDomain(value)
      if (value !== '') {
        this.props.fetchCommunityExists(value)
      }
    }
    this.setState({
      edits: {
        ...this.state.edits,
        [name]: value,
        changed: true
      }
    })
  }

  onEnter = event => {
    if (event.key === 'Enter') {
      this.errorCheckAndSubmit()
    }
  }

  submit = () => {
    const { name, email, communityName, communityDomain, communityNetworkId } = this.state.edits
    this.state.edits.changed && this.props.updateUserSettings({
      name,
      email
    })
    this.props.createCommunity(
      // communityPrivacy,
      communityName,
      communityDomain,
      communityNetworkId
    )
      .then(({ error }) => {
        if (error) {
          this.setState({
            error: 'There was an error, please try again.'
          })
        } else {
          this.props.goToCommunity(`${communityDomain}`)
        }
      })
  }

  errorCheckAndSubmit = () => {
    const { name, email, communityName, communityDomain } = this.state.edits
    if (name === '' || email === '' || communityName === '' || communityDomain === '') {
      this.setState({
        error: 'Please fill in each field.'
      })
    } else if (this.props.communityDomainExists) {
      this.setState({
        error: 'This URL already exists. Try another.'
      })
    } else if (!slugValidatorRegex.test(removeUrlFromDomain(communityDomain))) {
      formatDomainWithUrl(communityDomain)
      this.setState({
        error: invalidSlugMessage
      })
    } else {
      this.submit()
    }
  }

  componentWillMount = () => {
    const { communityPrivacy } = this.props
    const privacyOption = find(privacyOptions, { label: communityPrivacy })
    const selectedCommunityPrivacy = get('label', privacyOption) // set to Private by default
    this.setState({
      edits: {
        name: get('name', this.props.currentUser) || '',
        email: get('email', this.props.currentUser) || '',
        communityName: get('communityName', this.props) || '',
        communityDomain: get('communityDomain', this.props) || '',
        communityNetworkId: get('communityNetworkId', this.props) || '',
        communityPrivacy: selectedCommunityPrivacy
      }
    })
  }

  formatDomainWithUrl (communityDomain) {
    if (!communityDomain) return null
    let formattedDomain = communityDomain.replace('hylo.com/c/', '').replace('hylo.com/c', '')
    if (formattedDomain !== '') {
      formattedDomain = 'hylo.com/c/' + formattedDomain
    }
    return formattedDomain
  }

  removeUrlFromDomain (communityDomain) {
    return communityDomain.replace('hylo.com/c/', '')
  }

  render () {
    const { networkName } = this.props

    return <div styleName='flex-wrapper'>
      <ModalSidebar
        imageUrl={groovingAxolotl}
        onClick={this.props.goHome}
        theme={sidebarTheme}
        header='Everything looking good?'
        body='You can always come back and change your details at any time'
      />
      <div styleName='panel'>
        <div>
          <span styleName='step-count'>STEP 3/3</span>
        </div>
        <div styleName='center'>
          <div styleName='logo center' style={bgImageStyle(hyloNameWhiteBackground)} />
        </div>
        <div styleName='center-review'>
          <ReviewTextInput
            label={'Your Name'}
            value={this.state.edits.name || ''}
            readOnly={this.state.readOnly.name}
            editHandler={(event) => this.editHandler('name')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'name')}
          />
          <ReviewTextInput
            label={'Your Email'}
            value={this.state.edits.email || ''}
            readOnly={this.state.readOnly.email}
            editHandler={() => this.editHandler('email')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'email')}
          />
          <ReviewTextInput
            label={'Community Name'}
            value={this.state.edits.communityName || ''}
            readOnly={this.state.readOnly.communityName}
            editHandler={() => this.editHandler('communityName')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'communityName')}
          />
          <ReviewTextInput
            label={'URL'}
            value={formatDomainWithUrl(this.state.edits.communityDomain) || ''}
            readOnly={this.state.readOnly.communityDomain}
            editHandler={() => this.editHandler('communityDomain')}
            onEnter={this.onEnter}
            onChange={(e) => this.handleInputChange(e, 'communityDomain')}
          />
          {networkName && <ReviewTextInput
            label={'Network'}
            value={networkName}
            readOnly
          />}
          {/* <ReviewTextInput */}
          {/* label={'Privacy'} */}
          {/* value={this.state.edits.communityPrivacy} */}
          {/* onEnter={this.onEnter} */}
          {/* editHandler={() => this.props.goToPrivacyStep()} */}
          {/* /> */}

        </div>
        { this.state.error && <span styleName='review-arrow-up' /> }
        { this.state.error && <span styleName='review-error'>{this.state.error}</span>}
      </div>
      <ModalFooter
        submit={this.errorCheckAndSubmit}
        previous={this.previous}
        showPrevious={false}
        continueText={'Finish Up'}
      />
    </div>
  }
}

export function ReviewTextInput ({ label, value, editHandler, onChange, readOnly = true, inputRef }) {
  return <div styleName='review-input-text-row'>
    <div styleName='review-input-text-label'>
      <span>{label}</span>
    </div>
    <div styleName='review-input-text'>
      <TextInput
        type='text'
        name='community-name'
        value={value}
        theme={inputTheme}
        readOnly={readOnly}
        noClearButton
        onChange={onChange}
        inputRef={inputRef}
      />
    </div>
    {editHandler && <div styleName='review-input-edit'>
      <span styleName='edit-button' onClick={editHandler}>Edit</span>
    </div>}
  </div>
}

const inputTheme = {
  inputStyle: 'modal-input partial',
  wrapperStyle: 'center'
}

const sidebarTheme = {
  sidebarHeader: 'sidebar-header-full-page',
  sidebarText: 'gray-text sidebar-text-full-page'
}

const privacyOptions = [
  { id: '0', label: 'public' },
  { id: '1', label: 'private' },
  { id: '2', label: 'unlisted' }
]
