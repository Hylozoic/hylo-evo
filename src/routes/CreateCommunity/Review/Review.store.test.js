import { createCommunity } from './Review.store'

describe('createCommunity', () => {
  describe('without networkId', () => {
    it('omits networkId from variables', () => {
      const action = createCommunity('thename', 'theslug')
      expect(action).toMatchSnapshot()
      expect(action.graphql.variables.data.networkId).toEqual(undefined)
    })
  })

  describe('with networkId', () => {
    it('includes networkId in variables', () => {
      const networkId = 123
      const action = createCommunity('thename', 'theslug', networkId)
      expect(action).toMatchSnapshot()
      expect(action.graphql.variables.data.networkId).toEqual(networkId)
    })
  })
})
