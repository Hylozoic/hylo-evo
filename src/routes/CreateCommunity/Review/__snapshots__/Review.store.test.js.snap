// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCommunity with networkId includes networkId in variables 1`] = `
Object {
  "graphql": Object {
    "query": "mutation ($data: CommunityInput) {
        createCommunity(data: $data) {
          id
          hasModeratorRole
          community {
            id
            name
            slug
            network {
              id
            }
          }
        }
      }
      ",
    "variables": Object {
      "data": Object {
        "name": "thename",
        "networkId": 123,
        "slug": "theslug",
      },
    },
  },
  "meta": Object {
    "extractModel": "Membership",
    "name": "thename",
    "networkId": 123,
    "slug": "theslug",
  },
  "type": "Review/CREATE_COMMUNITY",
}
`;

exports[`createCommunity without networkId omits networkId from variables 1`] = `
Object {
  "graphql": Object {
    "query": "mutation ($data: CommunityInput) {
        createCommunity(data: $data) {
          id
          hasModeratorRole
          community {
            id
            name
            slug
            network {
              id
            }
          }
        }
      }
      ",
    "variables": Object {
      "data": Object {
        "name": "thename",
        "slug": "theslug",
      },
    },
  },
  "meta": Object {
    "extractModel": "Membership",
    "name": "thename",
    "slug": "theslug",
  },
  "type": "Review/CREATE_COMMUNITY",
}
`;
